### PR TITLE
KVStore: Reduce the memory before and after a large txn is committed

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #9722

Problem Summary:

### What is changed and how it works?

There are two ideas:
1. The first idea is to reduce the memory after a large txn is commmitted. The mechanism is introduced in https://github.com/pingcap/tidb-engine-ext/pull/402. Merged in https://github.com/pingcap/tiflash/pull/9721
2. The second idea is to reduce the memory before a large txn is committed. The idea is based on the observation that the `DecodedLockCFValue` has the size of 128 bytes. Given a txn that updates 15m rows, the memory cost is 1.8G

The idea is that we won't store the built(parsed) DecodedLockCFValue if the lock belongs to a large txn. So everytime we scan the locks of large txns, we will parse it again.

```commit-message
Avoid store the built DecodedLockCFValue
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/KVStore/MultiRaft/RaftCommands.cpp [hunk 1]
```diff
@@ -452,9 +452,17 @@ std::pair<EngineStoreApplyRes, DM::WriteResult> Region::handleWriteRaftCmd(
                     handle_by_index_func(i);
             }
         }
-        cache_written_size = dataSize() - ori_cache_size;
         approx_mem_cache_rows += cmd_write_cf_cnt;
-        approx_mem_cache_bytes += cache_written_size;
+        auto current_size = dataSize();
+        if (current_size > ori_cache_size)
```
Comment (CalvinNeo) [inline dbms/src/Storages/KVStore/MultiRaft/RaftCommands.cpp:457]: This size is not so accurate, need later investigating.

File: dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.cpp [hunk 2]
```diff
@@ -189,7 +218,22 @@ std::unique_ptr<kvrpcpb::LockInfo> DecodedLockCFValue::intoLockInfo() const
...
+kvrpcpb::Op DecodedLockCFValue::getLockType() const
+{
+    kvrpcpb::Op x;
+    withInner([&](const Inner & in) { x = in.lock_type; });
+    return x;
+}
+
+UInt64 DecodedLockCFValue::getLockVersion() const
+{
+    auto x = 0;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.cpp:234]:
```suggestion
    UInt64 x = 0;
```

File: dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.cpp [hunk 3]
```diff
@@ -23,9 +23,11 @@ namespace RecordKVFormat
 {
 
 // https://github.com/tikv/tikv/blob/master/components/txn_types/src/lock.rs
-inline void decodeLockCfValue(DecodedLockCFValue & res)
+[[nodiscard]] DecodedLockCFValue::Inner * decodeLockCfValue(const DecodedLockCFValue & decoded)
 {
-    const TiKVValue & value = *res.val;
+    auto * inner = new DecodedLockCFValue::Inner{};
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.cpp:28]: OK

**Comment by CalvinNeo** [comment]
------------------------------------------------------------
Benchmark                  Time             CPU   Iterations
------------------------------------------------------------
parseTest                126 ns          125 ns      5612257
parseTest                129 ns          129 ns      5612257
parseTest_mean           127 ns          127 ns            2
parseTest_median         127 ns          127 ns            2
parseTest_stddev        2.56 ns         2.56 ns            2
parseTest_cv            2.01 %          2.02 %             2

File: dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp [hunk 4]
```diff
@@ -275,66 +275,65 @@ std::variant<RegionDataReadInfoList, RegionException::RegionReadStatus, LockInfo
     bool need_data_value)
 {
     RegionDataReadInfoList data_list_read;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp:277]:
```suggestion
```

File: dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp [hunk 5]
```diff
@@ -275,66 +275,65 @@ std::variant<RegionDataReadInfoList, RegionException::RegionReadStatus, LockInfo
...
-        {
-            /// Check if there are any lock should be resolved, if so, throw LockException.
-            lock_value
-                = scanner.getLockInfo(RegionLockReadQuery{.read_tso = start_ts, .bypass_lock_ts = bypass_lock_ts});
-        }
+    /// If there is no lock, leave scope of region scanner and raise LockException.
+    /// Read raw KVs from region cache.
+    if (!lock_info)
+    {
+        // Shortcut for empty region.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp:320]:
```suggestion
        RegionDataReadInfoList data_list_read;
        // Shortcut for empty region.
```

File: dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp [hunk 6]
```diff
@@ -275,66 +275,65 @@ std::variant<RegionDataReadInfoList, RegionException::RegionReadStatus, LockInfo
...
-
-            // Tiny optimization for queries that need only handle, tso, delmark.
-            do
-            {
-                data_list_read.emplace_back(scanner.next());
-            } while (scanner.hasNext());
-        }
+            data_list_read.emplace_back(scanner.next());
+        } while (scanner.hasNext());
+    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp:332]:
```suggestion
        } while (scanner.hasNext());
        return data_list_read;
    }
```

File: dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp [hunk 7]
```diff
@@ -275,66 +275,65 @@ std::variant<RegionDataReadInfoList, RegionException::RegionReadStatus, LockInfo
...
+    else
+    {
+        return lock_info;
     }
-
-    if (lock_value)
-        return lock_value->intoLockInfo();
 
     return data_list_read;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp:338]:
```suggestion
```

File: dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp [hunk 8]
```diff
@@ -275,66 +275,65 @@ std::variant<RegionDataReadInfoList, RegionException::RegionReadStatus, LockInfo
...
-                data_list_read.emplace_back(scanner.next());
-            } while (scanner.hasNext());
-        }
+            data_list_read.emplace_back(scanner.next());
+        } while (scanner.hasNext());
+    }
+    else
+    {
+        return lock_info;
     }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/Decode/PartitionStreams.cpp:336]:
```suggestion
    return lock_info;
```

File: dbms/src/Storages/KVStore/MultiRaft/RegionData.cpp [hunk 9]
```diff
@@ -222,22 +223,37 @@ DecodedLockCFValuePtr RegionData::getLockInfo(const RegionLockReadQuery & query)
...
+            {
+                should_continue = true;
+                return;
+            }
+            if (in.min_commit_ts > query.read_tso)
             {
-                GET_METRIC(tiflash_raft_read_index_events_count, type_bypass_lock).Increment();
-                continue;
+                should_continue = true;
+                return;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/MultiRaft/RegionData.cpp:239]: `bool withInner(std::function<bool(const Inner &)> f) const;` may be better?

Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/MultiRaft/RegionData.cpp:239]: Similar, I can move it into the DecodedLockCfValue

File: dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.h [hunk 10]
```diff
@@ -23,25 +23,36 @@ namespace DB::RecordKVFormat
...
-    bool use_async_commit{0};
-    UInt64 min_commit_ts{0};
-    std::string_view secondaries;
-    std::string_view primary_lock;
-    bool is_txn_file{0};
-    // For large txn, generation is not zero.
-    UInt64 generation{0};
+
+private:
+    friend DecodedLockCFValue::Inner * decodeLockCfValue(const DecodedLockCFValue & res);
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Storages/KVStore/TiKVHelpers/DecodedLockCFValue.h]: Fixed

**Review by Lloyd-Pottiger — APPROVED** [review]

**Pull request merged**